src/_pytest/main.py

```diff
@@ -249,7 +253,11 @@ def pytest_cmdline_main(config):
     return wrap_session(config, _main)
 
 
-def _main(config, session):
+def _main(
+    config: Config, session: "Session"
+) -> Optional[
+    Union["Literal"[ExitCode.TESTS_FAILED], "Literal"[ExitCode.NO_TESTS_COLLECTED]]
```

Comment:
Are you sure `_main` will only ever return these two error codes? If it is put in the type annotation, then once they're published, it becomes an API guarantee.

Reply:
That's what is used in that function.  Also `_main` is not really public API, is it?

Reply:
It's what used in the function but it's better to keep our options open IMO.

IIUC, eventually this return value is funneled to the `main` function in `_pytest/config/__init__.py` which has return type `Union[int, _pytest.main.ExitCode]`. My suggestion is to use the same type here.

Reply:
Sorry, I fail to see why.  If a function returns only `1`, or `2`, it should/could have `Literal[1, 2]`, even when used from another function that returns `int`, no?

Reply:
This is really not very important but I'll try to explain my view.

I think restricting the type in such a way is not beneficial. The function is an (indirect) implementation of a hook and the only restriction that needs to apply to it is that it implements the hook specification.

Unless this is an actual API guarantee that should not be changed without it being a breaking change, the fact that it currently returns these two values is just an incidental implementation detail. If someone wants to say add a new error code, then they 1) might be unsure if they're allowed to change the type, and 2) need to "uselessly" adjust the type.

Reply:
I've really seen it more like being able to see from the signature already, what it might "raise".
But I do not have a strong opinion on it, and certainly trust you more with regard to type hinting.. :)
Amended.

Reply:
Thanks for indulging me :)

Reply:
@bluetech just for understanding: would you type a function always returning True also with `bool` then?  (https://github.com/blueyed/pytest/blob/70481429f8236141e3a39098082b254f4a1cc017/src/_pytest/python.py#L172)

My general understanding/idea with that was that we could later tell that a certain hook, when used, would always skip later ones.  This is somehow indicated with `-> bool` also (not `Optional[bool]`, but following your argument it should be typed like the hookspec, which is `Optional[…]`/`Optional[Any]` then for `firstresult=True`.

Reply:
I'd tend to use the same type as the hookspec. I see that fact that the impl currently skips (returns True) is a runtime property, which might change. But if it's really fundamental, as in we want to "enforce" that this particular impl always skips, then making it `-> True` does make sense.